### PR TITLE
ENH: enable specifying custom conda channels + create env via environment.yml

### PR DESCRIPTION
**Pull request by tylerjereddy**
- This PR adds the basic functionality to specify a list of custom conda channels using a new key in the JSON config file; the associated feature request issue is #405 
- This actually looks fairly clean to me (and works for me), but there are probably some things that the core devs will know that I haven't thought of
- I suspect the way the channels are added may need slight modification -- installing packages is done in a single call to conda for performance reasons, and we may wish to do the same for adding channels (I currently loop over the input channels as you can see)
- need for tests? -- make sure this works when adding a single vs. multiple custom channels? mock the call to those channels? how much of a pain would this be to do? Maybe there's something simpler that can be done test-wise.

### CODE REVIEW

**Comment by tylerjereddy** [comment]
AppVeyor succeeded; all travis tests pass except for Python 3.3, which I don't understand.

**Comment by kain88-de** [comment]
> need for tests? -- make sure this works when adding a single vs. multiple custom channels? mock the call to those channels? how much of a pain would this be to do? Maybe there's something simpler that can be done test-wise.

You can check the created `~/.condarc` yaml file. As long channels are added you are fine. This also saves you from writing a mock.



> I suspect the way the channels are added may need slight modification -- installing packages is done in a single call to conda for performance reasons, and we may wish to do the same for adding channels (I currently loop over the input channels as you can see)

You can also directly modify `~/.condarc` but the calls to add a channel are usually fast and I do not expect that  people want to add 100 different channels.


Otherwise I like this addition since conda-forge packages a lot of dependencies for projects that are not available on the default channel.

**Comment by kain88-de** [comment]
@tylerjereddy this will always keep the default channel. I never removed that one myself but for others it might be useful to allow removing it and only use custom channels

**Comment by tylerjereddy** [comment]
@kain88-de The latest commit includes a draft unit test for verifying the addition of channels via probing the output of `conda info` -- which I think just probes the `~/.condarc` data anyway.

One major drawback of this approach is that it touches the actual `~/.condarc` on a users' machine if they run the unit test suite. Can we avoid this?

I think I also overestimated the sophistication of what is going on--although my unit test creates two separate environments for two versions of Python, all that really seems to happen is that global conda settings on a given machine are modified by asv and then used for those environments. 

I think I was originally envisioning more isolation for the environments, but if my interpretation is correct, I'm not even sure we need to create the environments for this unit test (though it may not hurt) -- it seems the main thing is just to check that the machine conda settings have been adjusted appropriately.

**Comment by kain88-de** [comment]
> One major drawback of this approach is that it touches the actual ~/.condarc on a users' machine if they run the unit test suite. Can we avoid this?

You can be specifying which  channel a package should be installed from. Then you can call `conda install -c custom_channel package`. But that requires to change the way conda packages are installed and how to specify a package.

**Comment by pv** [comment]
The config setting looks OK on a quick look. Add the empty list value in 
https://github.com/spacetelescope/asv/blob/master/asv/config.py#L21 and 
remove the try-except check.

Changing ~/.condarc cannot be done, so some other approach should be 
used to inform conda to what to use.

**Comment by kain88-de** [comment]
Conda has a feature to [share environments ](https://conda.io/docs/using/envs.html#share-an-environment). What would prohibit us from using that feature directly? It allows to specify the channels I'm using as well as the packages with optional version tracking.  Installation can still be done in a single step here.


Here is the environment file for my own mdanalysis environment.

```
name: mda
channels:
- kain88-de
- bioconda
- conda-forge
- defaults
dependencies:
- biopython=1.70=np112py36_0
- joblib=0.9.3=py36_0
- typing=3.5.2.2=py36_0
- alabaster=0.7.10=py36_0
- apipkg=1.4=py36_0
- asn1crypto=0.22.0=py36_0
- babel=2.3.4=py36_0
- blas=1.1=openblas
- ca-certificates=2017.7.27.1=0
- certifi=2017.7.27.1=py36_0
- cffi=1.10.0=py36_0
- chardet=3.0.2=py36_1
- coverage=4.4.1=py36_0
- cryptography=1.9=py36_0
- curl=7.54.1=0
- cycler=0.10.0=py36_0
- cython=0.26=py36_0
- dbus=1.10.10=3
- decorator=4.1.2=py36_0
- docutils=0.13.1=py36_0
- execnet=1.4.1=py36_0
- expat=2.2.1=0
- fontconfig=2.12.1=4
- freetype=2.7=1
- gettext=0.19.7=1
- glib=2.51.4=0
- griddataformats=0.3.3=py36_0
- gst-plugins-base=1.8.0=0
- gstreamer=1.8.0=2
- hdf4=4.2.12=0
- hdf5=1.8.18=0
- hypothesis=3.11.6=py36_0
- icu=58.1=1
- idna=2.5=py36_0
- imagesize=0.7.1=py36_0
- ipython=6.1.0=py36_0
- ipython_genutils=0.2.0=py36_0
- jedi=0.10.2=py36_0
- jinja2=2.9.5=py36_0
- jpeg=9b=0
- krb5=1.14.2=0
- libffi=3.2.1=3
- libiconv=1.14=4
- libnetcdf=4.4.1.1=5
- libpng=1.6.28=0
- libssh2=1.8.0=1
- libtiff=4.0.6=7
- libxcb=1.12=1
- libxml2=2.9.4=4
- markupsafe=1.0=py36_0
- matplotlib=2.0.2=py36_1
- mmtf-python=1.0.7=py36_0
- msgpack-python=0.4.8=py36_0
- ncurses=5.9=10
- netcdf4=1.2.9=np112py36_0
- networkx=1.11=py36_0
- nose=1.3.7=py36_2
- numpy=1.12.1=py36_blas_openblas_200
- olefile=0.44=py36_0
- openblas=0.2.19=2
- openssl=1.0.2l=0
- pandas=0.20.3=py36_1
- patsy=0.4.1=py36_0
- pcre=8.39=0
- pexpect=4.2.1=py36_0
- pickleshare=0.7.3=py36_0
- pillow=4.2.1=py36_0
- pip=9.0.1=py36_0
- prompt_toolkit=1.0.15=py36_0
- ptyprocess=0.5.2=py36_0
- py=1.4.34=py36_0
- pycparser=2.18=py36_0
- pygments=2.2.0=py36_0
- pyopenssl=16.2.0=py36_0
- pyparsing=2.2.0=py36_0
- pyqt=5.6.0=py36_4
- pysocks=1.6.7=py36_0
- pytest=3.2.0=py36_0
- pytest-cov=2.5.1=py36_0
- pytest-xdist=1.18.2=py36_0
- python=3.6.2=0
- python-dateutil=2.6.1=py36_0
- pytz=2017.2=py36_0
- qt=5.6.2=3
- readline=6.2=0
- reportlab=3.4.0=py36_0
- requests=2.18.3=py36_0
- scikit-learn=0.18.2=np112py36_blas_openblas_200
- scipy=0.19.1=py36_blas_openblas_201
- seaborn=0.8.0=py36_0
- setuptools=36.2.2=py36_0
- simplegeneric=0.8.1=py36_0
- sip=4.18=py36_1
- six=1.10.0=py36_1
- snowballstemmer=1.2.1=py36_0
- sphinx=1.6.3=py36_0
- sphinxcontrib-websupport=1.0.1=py36_0
- sqlite=3.13.0=1
- statsmodels=0.8.0=np112py36_0
- tk=8.5.19=2
- tornado=4.5.1=py36_0
- traitlets=4.3.2=py36_0
- urllib3=1.21.1=py36_1
- wcwidth=0.1.7=py36_0
- wheel=0.29.0=py36_0
- xorg-libxau=1.0.8=3
- xorg-libxdmcp=1.1.2=3
- xz=5.2.2=0
- zlib=1.2.11=0
- libgfortran=3.0.0=1
- llvmlite=0.19.0=py36_0
- numba=0.34.0=np112py36_0
- pip:
  - ipython-genutils==0.2.0
  - prompt-toolkit==1.0.15
prefix: /home/max/anaconda3/envs/mda
```

**Comment by tylerjereddy** [comment]
It looks like sharing environments requires access to an `environment.yml` file and the docs even suggest emailing that file to another person that wants to use the same environment, which doesn't seem like a suitable automated / general approach for environment generation in `asv`.

If `conda install -c channel_1 channel_2 channel_3 package`  were possible, that might be a clean approach, since you could just use the custom channels in the specified priority order without modifying condarc. 

However, it seems that you can only specify a single channel this way (i.e., the one you want to use for the specified package), so the `-c` flag approach (as I think @kain88-de suggests) would appear to require specification of a more complex dependency matrix that associates every required package with its cognate channel; the asv JSON file may get pretty gruesome with a dictionary-style association like this.

**Comment by kain88-de** [comment]
> It looks like sharing environments requires access to an environment.yml file and the docs even suggest emailing that file to another person that wants to use the same environment, which doesn't seem like a suitable automated / general approach for environment generation in asv.

It should be possible to place this into the asv config file. I would also expect I can generate such a file from the current asv config files. Also with your suggested additions to specify a channel.

**Comment by pv** [comment]
You should be able generate a temporary environment.yml that is used in 
the environment creation.

**Comment by tylerjereddy** [comment]
Ok, so I think I understand the proposed route forward -- programmatically generate `environment.yml` from the asv config, including channels in the specified priority order.

A user could specify `nodefaults` as a channel to exclude the default channel without us having to add another JSON flag for that functionality, as documented for the `.yml` conda file.

**Comment by kain88-de** [comment]
The suggested solution might break existing workflows for people who implicitly depend on the conda channels they define in their own `.condarc`. Personally I like using a environment file better because it leads to better reproducible  environments. So this change should be mentioned more explicit in the release notes.

**Comment by tylerjereddy** [comment]
I started hacking away at the usage of an `environment.yml` file in the latest commit. I only get about three test failures locally, but there is definitely still a little more design / thought needed I think.

A common thing I see is this: `"CondaValueError: Value error: prefix already exists: /Users/treddy/miniconda2/envs/conda-py2.7-colorama0.3.6-six"`

And indeed, `conda env list` does get polluted with the new environments that are created during the tests. I think there's a flag to force the creation if the env already exists, but probably not the way to go.

Should we try to clean up the environment names after the tests /benchmarks run? I'm slightly concerned about the fragility of this approach--i.e., that the cleanup may not happen if an end-user has a test failure before it can happen & indeed the simple fact of modifying a users' `conda env list` is perhaps problematic in general (is there a nice trick to avoid this completely?).

Perhaps specifying the full path to the environment prefix for `conda env create` & using a temp dir of some sort could prevent polluting i.e. the local env list?

**Comment by pv** [comment]
It's looks like it's creating the envs to the wrong place, they should go
to self._path. Conda create had a -p option. Can that be used with
environment.html?

29.8.2017 1.47 "Tyler Reddy" <notifications@github.com> kirjoitti:

I started hacking away at the usage of an environment.yml file in the
latest commit. I only get about three test failures locally, but there is
definitely still a little more design / thought needed I think.

A common thing I see is this: "CondaValueError: Value error: prefix already
exists: /Users/treddy/miniconda2/envs/conda-py2.7-colorama0.3.6-six"

And indeed, conda env list does get polluted with the new environments that
are created during the tests. I think there's a flag to force the creation
if the env already exists, but probably not the way to go.

Should we try to clean up the environment names after the tests /benchmarks
run? I'm slightly concerned about the fragility of this approach--i.e.,
that the cleanup may not happen if an end-user has a test failure before it
can happen & indeed the simple fact of modifying a users' conda env list is
perhaps problematic in general (is there a nice trick to avoid this
completely?).

Perhaps specifying the full path to the environment prefix for conda env
create & using a temp dir of some sort could prevent polluting i.e. the
local env list?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/spacetelescope/asv/pull/539#issuecomment-325504635>,
or mute
the thread

**Comment by tylerjereddy** [comment]
Looks like it can. I'm down to two (local) test failures now: test_conda_pip_install & test_continuous. Hopefully I can take a closer look at those two issues soon.

`conda env list` no longer gets polluted by running the test suite now

**Comment by tylerjereddy** [comment]
I was able to resolve one test failure related to an [open conda bug](https://github.com/conda/conda/issues/4985) where pip needs to be specified as a dependency within `environment.yml`. Actually, the latter conda bug is labelled as an easy fix, but might be a bit much at this stage to go even farther upstream fixing issues.

Anyway, that's good, we're a tiny bit closer. I note that the total number of failures in the test suite does differ a bit for the same commit hash on different operating systems sometimes, at least for this PR branch.

**Comment by tylerjereddy** [comment]
I've patched all 7 test failures on a linux machine with the latest changes -- let's see how the continuous integration works out now.

**Comment by tylerjereddy** [comment]
Both Travis and AppVeyor builds are passing. At this stage, I think at the very least we need:

- cleanup of affected function names and comments / docs -- in a few places I've repurposed a function without changing its name to reflect this, and some comments probably need adjustment as well
- the original unit test I added is probably out of date now -- an appropriate / updated version should be written and then the test can be reactivated
- we might want to think about one or two more test cases; indeed, there is currently a modest drop in test coverage

Still, I think the basic functionality is here now.

**Comment by tylerjereddy** [comment]
Although the tests pass, as I try to write a specific test for channel addition, I'm noticing that `environment.yml` does not show up in `self._path` when I check on my machine after local tests. I can see `asv-env-info.json` there & the code to write the two files looks relatively similar to me, but there must be something tricky about writing files to the environment.

Furthermore, the environment creation command must be working for the tests to pass, and I can open the `environment.yml` file from within `plugins/conda.py` after creating it to print the contents, but not from within the test suite or when manually cd'ing to the directory in /tmp.

**Comment by tylerjereddy** [comment]
I've updated the new tests to reflect the desired / enforced behavior. I note that `environment.yml` is actually obliterated from the env by the `--force` flag I used in `conda env create` to avoid certain test error scenarios where a conda env already existed. This is mostly a side-effect of the behavior that allows the unit tests to all pass & removing the `yml` file from the env is not necessarily something I was aiming to do. We could possibly take measures to restore it after if this is considered important (it may be useful residual info to have in the env folder).

One of the new unit tests verifies that the order of channels that is specified for `environment.yml` is respected when installing packages. I note that actually parsing which channels are added by `asv` to the env was rather tricky. `conda info` doesn't seem applicable (indeed I think we actually want to avoid modifying this data just as we want to avoid touching `condarc`), and I actually ended up parsing the json data from `conda list` within the env to find the source channels used to install those packages. That's the best approach I've found so far.

Let's see what the CI says now.

**Comment by pv** [comment]
Note that the method names need to be changed --- 
`_install_requirements` does not actually install requirements, and 
`_add_channels` does not add channels. I'd include the `_add_channels` 
in the body of `_setup` without a separate method. What 
`_install_requirements` does sounds more like `_get_requirements`.

`mkdir_p` should not be needed, the path is guaranteed to already exist 
when `_setup` is called. If it's really needed, it should go in util.py

You can also write some of the file generation stuff simpler, e.g.,
```
+            env_file.write('dependencies:\n'
+                           '   - python={0}\n'
+                           '   - wheel\n'
+                           '   - pip\n'.format(self._python))
```
Note how strings can be split to multiple lines and that parenthesis is 
not needed before .format.

Instead of using `create --force` wouldn't it be better to create the 
environment.yml as a temporary file? Like so:
```
with tempfile.NamedTemporaryFile() as env_file:
     env_file.write(...)
     ...
     env_file.flush()
     util.check_output([conda, 'env', 'create', '-f', env_file.name] + ...)

```

**Comment by tylerjereddy** [comment]
Ok, I think I've addressed most of the comments. I suspect there will still be an issue with the AppVeyor test--not exactly sure what the problem there is yet, but perhaps something quite simple involving windows paths.

**Comment by tylerjereddy** [comment]
- Travis stuff mostly looks ok -- one strange failure for Python 3.5 only -- is that real?
- I adopted the proposed `NamedTemporaryFile`  solution, but I don't understand how you are planning to overcome the platform-specific limitation here -- `conda env create` won't be able to access the file name for opening a second time on Windows [as the std lib docs indicate](https://docs.python.org/2/library/tempfile.html#tempfile.NamedTemporaryFile)

There's an [open bug report in for CPython](https://bugs.python.org/issue14243) for this portability issue; looks like it is targeted for Python 3.7 maybe. Basically, you can open the tempfile a second time in the context manager on unix but not on Windows.

@pv my suggestion is that we use `NamedTemporaryFile(mode='w', delete=False)` so that the tempfile is not actually deleted after the context manager control flow exits, and then we move the `conda env create` call outside of the context manager. This is far less Pythonic, and kind of defeats the purpose of using a tempfile, but it does circumvent the issue, and eventually we can use the more Pythonic approach when this gets patched in CPython and platform portability is restored.

**Comment by tylerjereddy** [comment]
We may also have to switch back to using `--force` with conda as well, because we won't be able to leverage the cleanup of the tempfile if we use delete=False.

**Comment by tylerjereddy** [comment]
I've pushed the attempted windows compatibility changes despite the issues discussed above -- let's see what happens with the CI now.

**Comment by tylerjereddy** [comment]
So, less issues now:
- All Windows builds have only two failures, both for the same unit test because of `conda list -p env._path --json` failing on that platform for whatever reason (file path handling needs to be done more carefully in that unit test maybe?)
- only the Python 2.6 Travis test suite has a single failure; all others pass -- looks like an issue with pip wheel in Python 2.6

**Comment by tylerjereddy** [comment]
`util.check_output([conda, 'list'])` is sufficient to trigger the missing file failure on Windows. Now testing to see what path for `conda` is being used on Windows, but whatever the path, it is already clear that `os.path.normpath()` is not sufficient to make it work on that platform.

**Comment by tylerjereddy** [comment]
So, `env._path` exists as expected on the Windows / AppVeyor builds, but the `bin` subdirectory does not. I'm not sure if a symlink to the `conda` executable is being created within (another subdir of) `env._path` on Windows, but I'm pretty sure that the unix builds do create a symlink to conda executable somewhere (I remember seeing it when debugging).

There is a `conda-meta` entry in `env._path` on windows. I can take a peek in there.

My current suspicion is that `env.find_executable()` may need to be modified to point to the proper Windows executable for `conda`, because what it gives me now doesn't exist. There is a comment in the latter function referencing a standard conda layout, which is perhaps now no longer the case with my modifications, but it could also be something more confusing.

**Comment by tylerjereddy** [comment]
My debug work indeed indicates that there's a `conda` file (probably a symlink type thing) in the `Scripts` dir on Windows / AppVeyor when building an env. `asv` assumes that the executable for conda on windows will have an `exe` extension in the env and reside in either the env root or the `Scripts` subdirectory, based on the code in `find_executable`. Since this is not the case in the current AppVeyor builds, `asv` falls back to assuming that we have a unix configuration with a `bin` path instead, but it is in fact a Windows build that appears to have a `conda` executable without an `exe` extension.

So, I'll aim to try to patch `find_executable` to handle this scenario. I'm not clear on whether this is genuinely related to anything I've changed rather than simply the current tests don't probe this issue, but I think the logical fix is becoming clearer now.

**Comment by pv** [comment]
Maybe related: https://github.com/spacetelescope/asv/pull/525

**Comment by pv** [comment]
Maybe the minimal thing would be to add a `paths=None` argument that 
takes a list of paths [with default value 
os.environ['PATH'].split(os.pathsep)] to util.which so that the thing is 
done only in a single place.

**Comment by tylerjereddy** [comment]
So you're suggesting it would be more appropriate to use `util.which` than `find_executable` in my unit test to probe for conda, and to modify util.which for increased flexibility if needed? `find_executable` doesn't seem to directly interact with `util.which`, so I'm guessing that changing the latter won't affect the current test.

**Comment by pv** [comment]
find_executable should be implemented using util.which(executable, 
paths=[...]), so that the logic of appending exe, bat, etc. on Windows 
is not duplicated in many places.

**Comment by pv** [comment]
Note that it may be faster to grab a windows VM and debug locally, than triaging it via appveyor:
https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/
https://developer.microsoft.com/en-us/windows/downloads/virtual-machines

**Comment by tylerjereddy** [comment]
I noticed that on a Windows machine `test_matrix_environments()` can be forced to use the `python.exe` within `self._path` with some minor changes. However, even with this adjustment, the libraries installed match the system rather than the environment if probed using the above `python.exe`. Further evidence that the system (or root / other conda) python is effectively being used by this test include variabilities in the library causing the version check failure when comparing AppVeyor test builds with builds on my Windows 10 machine / powershell.

That said, there is evidence that the original `conda env create` was done correctly because I can manually go into the pertinent `tmp` path on Windows and probe the expected library versions using `conda list -p (self._path)`.

What does this mean? Should there be a separate `python.exe` in the `Scripts` subfolder on windows? I'm only seeing one in the env `self._path`, and while the *version* of Python that is used does match the matrix, the version of the conda-installed libraries does not reflect the dependency matrix of the env (and probing the paths of these python libraries also doesn't match the env).

I think I'm slightly confused by the raw `python -c` calls in this unit test. I'm assuming we don't need to actually 'activate env' via conda  (there is no such call in the source anywhere) on the basis that the Python executable that is found is the one that would normally be 'activated' in any case, but I must be missing a little something here. This unit test is also rather prone to accidental passing since the system install of python could just happen to have dependency version matches to the env.

**Comment by tylerjereddy** [comment]
- all the tests are passing on all platforms except for a single fail on Python 2.6 -- that failure I don't really understand, but it is an unsupported version of Python anyway
- there are maybe one or two debug-related lines of code that should be removed, but the rest should be looking pretty solid now
- we could maybe add the unit tests from the issue @pv linked above and ensure that we handle those file types as well, although that could also be considered a separate issue which is only partly addressed here to obtain the desired functionality, which was really only supposed to be custom conda channel support

File: asv/util.py [hunk 1]
```diff
@@ -294,9 +292,17 @@ def which(filename):
 
     candidates = []
     for location in locations:
+        for alternate_ext in ['.exe', '.bat']:
```
Comment (pv) [inline asv/util.py]: This must be contained inside `if WIN` --- the stuff does not exist on other platforms.

File: asv/environment.py [hunk 2]
```diff
@@ -416,7 +416,9 @@ def check_presence(self):
 
         for executable in ['pip', 'python']:
             exe = self.find_executable(executable)
-            if not os.path.isfile(exe):
+            # do not count a system install of pip or python
+            # as evidence of an installed env
+            if not os.path.isfile(exe) or (self._path not in os.path.dirname(exe)):
```
Comment (pv) [inline asv/environment.py]: This check should be removed: the correct paths must be specified in find_executable.

File: asv/util.py [hunk 3]
```diff
@@ -272,20 +272,18 @@ def human_value(value, unit, err=None):
...
-    if WIN:
-        if not filename.endswith('.exe'):
-            filename = filename + '.exe'
-
     if os.path.sep in filename:
         locations = ['']
     else:
         locations = os.environ.get("PATH", "").split(os.pathsep)
+        if paths is not None:
+            locations = paths + locations
```
Comment (pv) [inline asv/util.py]: This should be:
```
if paths is not None:
    locations = paths
else:
    locations = os.environ ...
```

File: asv/plugins/conda.py [hunk 4]
```diff
@@ -85,30 +86,44 @@ def _matches(cls, python):
...
-            '--use-index-cache',
-            'python={0}'.format(self._python),
-            'pip'])
-
-        log.info("Installing requirements for {0}".format(self.name))
-        self._install_requirements(conda)
-
-    def _install_requirements(self, conda):
-        self.install('wheel')
+        # create a temporary environment.yml file
```
Comment (pv) [inline asv/plugins/conda.py]: Empty line before comment block, not after it.

File: asv/plugins/conda.py [hunk 5]
```diff
@@ -85,30 +86,44 @@ def _matches(cls, python):
...
+            env_file.write('name: {0}\n'
+                           'channels:\n'.format(self.name))
+            env_file.writelines(('   - %s\n' % ch for ch in self._conda_channels))
+            env_file.write('dependencies:\n'
+                           '   - python={0}\n'
+                           '   - wheel\n'
+                           '   - pip\n'.format(self._python))
+            # categorize & write dependencies based on pip vs. conda
+            reqs = self._get_requirements(conda)
+            # handle the case where there are multiple dependencies
```
Comment (pv) [inline asv/plugins/conda.py]: Empty line before comments block when at the same intendation as the preceding line.

File: asv/environment.py [hunk 6]
```diff
@@ -416,7 +416,9 @@ def check_presence(self):
 
         for executable in ['pip', 'python']:
             exe = self.find_executable(executable)
-            if not os.path.isfile(exe):
+            # do not count a system install of pip or python
```
Comment (pv) [inline asv/environment.py]: Empty line before comment blocks, when at the same intendation level as the preceding line.

File: asv/plugins/conda.py [hunk 7]
```diff
@@ -85,30 +86,44 @@ def _matches(cls, python):
 
     def _setup(self):
         try:
-            conda = util.which('conda')
+            conda = self.find_executable('conda')
```
Comment (pv) [inline asv/plugins/conda.py]: This must use `util.which`.
The environment is not initialized yet at this stage, so it should not try looking for the conda executable inside it.

File: asv/util.py [hunk 8]
```diff
@@ -272,20 +272,18 @@ def human_value(value, unit, err=None):
...
 
         if WIN:
             # On windows, an entry in %PATH% may be quoted
```
Comment (pv) [inline asv/util.py]: Also, this part can be run only when `paths is not None` because the `paths` list does not contain quoted paths.

File: asv/util.py (hunk 1 above)
Comment (pv) [reply asv/util.py]: Also, this is simpler to rewrite so that you build the list of possible filenames in a separate `if WIN: ... else: ...` block above: 
```
if WIN:
    ...
    filenames = [filename + ext for ext in ('.exe', '.bat', '')]
else:
    ...
    filenames = [filename]
...
for location in locations:
    for fn in filenames:
        candidate = os.path.join(location, fn)
        ...

**Comment by pv** [comment]
ok, fixed those myself now that I'm at it.

**Comment by pv** [comment]
Still needs:

- adding documentation to `docs/source/asv.conf.json.rst`
- adding the option to `asv/template/asv.conf.json`

**Comment by tylerjereddy** [comment]
Thanks for taking care of those! I pushed first draft documentation changes for the files you requested.

**Comment by pv** [comment]
Rebased and squashed, will still have to test on windows (separately, since appveyor doesn't seem to be on a cooperative mood)

**Comment by tylerjereddy** [comment]
The AppVeyor tests all pass now too. I note that on my Windows 10 machine I still get the 1 unit test failure I describe above involving `test_matrix_environments`.

I think my basic conclusion was that the conda env site package libs weren't always being used by the python executable in the env created by this test even though conda list -p self._path listed the correct dependency versions for the env.

Maybe it is ok since AppVeyor isn't reproducing this. For completeness, pytest traceback:

```
================================== FAILURES ===================================
__________________________ test_matrix_environments ___________________________

tmpdir = local('c:\\users\\treddy\\appdata\\local\\temp\\pytest-of-treddy\\pytest-45\\test_matrix_environments0')

    @pytest.mark.skipif(not ((HAS_PYTHON_27 and HAS_PYTHON_34) or HAS_CONDA),
                        reason="Requires Python 2.7 and 3.4")
    def test_matrix_environments(tmpdir):
        conf = config.Config()

        conf.env_dir = six.text_type(tmpdir.join("env"))

        conf.pythons = ["2.7", "3.4"]
        conf.matrix = {
            "six": ["1.10", None],
            "colorama": ["0.3.6", "0.3.7"]
        }
        environments = list(environment.get_environments(conf, None))

        assert len(environments) == 2 * 2 * 2

        # Only test the first two environments, since this is so time
        # consuming
        for env in environments[:2]:
            env.create()

            output = env.run(
                ['-c', 'import six, sys; sys.stdout.write(six.__version__)'],
                valid_return_codes=None)
            if 'six' in env._requirements:
>               assert output.startswith(six.text_type(env._requirements['six']))
E               AssertionError: assert False
E                +  where False = <built-in method startswith of unicode object at 0x000000000419F060>('1.10')
E                +    where <built-in method startswith of unicode object at 0x000000000419F060> = '1.11.0'.startswith
E                +    and   '1.10' = <type 'unicode'>('1.10')
E                +      where <type 'unicode'> = six.text_type

```

If I manually check:
```
PS C:\users\treddy\appdata\local\temp\pytest-of-treddy\pytest-45\test_matrix_environments0\env\d26ca092d5e1c69bb963927c9c813a5e> .\python.exe
Python 2.7.13 |Continuum Analytics, Inc.| (default, May 11 2017, 13:17:26) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import six
>>> six.__path__
[]
>>> import colorama
>>> colorama.__path__
['C:\\users\\treddy\\appdata\\local\\temp\\pytest-of-treddy\\pytest-45\\test_matrix_environments0\\env\\d26ca092d5e1c69bb963927c9c813a5e\\lib\\site-packages\\colorama']
>>> six.__version__
'1.11.0'
>>> colorama.__version__
'0.3.6'
>>> help(six)

Help on package six:

NAME
    six - Utilities for writing code that runs on Python 2 and 3

FILE
    c:\users\treddy\appdata\roaming\python\python27\site-packages\six-1.11.0-py2.7.egg\six.py
```

Looks like the env version of colorama is used, but not the case for six. The env packages are as expected though (including the version of `six` the unit test is looking for but doesn't find):
```
PS C:\users\treddy\appdata\local\temp\pytest-of-treddy\pytest-45\test_matrix_environments0\env\d26ca092d5e1c69bb963927c9c813a5e> conda list -p .
# packages in environment at C:\users\treddy\appdata\local\temp\pytest-of-treddy\pytest-45\test_matrix_environments0\env\d26ca092d5e1c69bb963927c9c813a5e:
#
certifi                   2016.2.28                py27_0
colorama                  0.3.6                    py27_0
pip                       9.0.1                    py27_1
python                    2.7.13                        1
setuptools                36.4.0                   py27_1
six                       1.10.0                   py27_0
vs2008_runtime            9.00.30729.5054               0
wheel                     0.29.0                   py27_0
wincertstore              0.2                      py27_0
```

Is Windows 10 newer than what is used in the current AppVeyor matrix? I suspect the actual issue is my usage of powershell instead of CMD --  I'm using a solution desribed in [an open conda issue](https://github.com/conda/conda/issues/626) for environment handling in this currently problematic scenario (powershell + Windows 10 + conda). So, perhaps we can consider that a conda issue or not critical.

In short, it seems that official conda versions don't support environment activation / control in powershell + Windows 10 at the moment, but there are some work-arounds described above. I'm not sure if that is really the source of the issue (I suppose if I test on CMD instead of powershell and everything passes that would be enlightening).

**Comment by pv** [comment]
The six on your machine looks like it comes from a `--user` installed 
package.

**Comment by pv** [comment]
This actually looks like a conda bug (would maybe be useful to report to 
them):
```
$ conda create -n foo python=3.6
...
$ source activate foo
$ python
Python 3.6.2 |Continuum Analytics, Inc.| (default, Jul 20 2017, 13:51:32)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
 >>> import sys
 >>> sys.path
['', '/home/pauli/.local/lib/conda/envs/foo/lib/python36.zip', 
'/home/pauli/.local/lib/conda/envs/foo/lib/python3.6', 
'/home/pauli/.local/lib/conda/envs/foo/lib/python3.6/lib-dynload', 
'/home/pauli/.local/lib/python3.6/site-packages', 
'/home/pauli/.local/lib/conda/envs/foo/lib/python3.6/site-packages']
#
# Note that /home/pauli/.local/lib/python3.6/site-packages
# appears in sys.path. This seems to only occur on Python 3.6.
#
```

**Comment by tylerjereddy** [comment]
Ok, I've opened the issue on the conda github (cross-ref'd above) after confirming on my machine that this doesn't happen with Python 2.7 or 3.5, so does indeed appear to be 3.6 only as you suggest. I offered to help them patch it if they want to guide me.

I didn't observe any other issues running the full `asv` testsuite on Win 10 / powershell.

**Comment by pv** [comment]
Thanks, merged!

**Comment by tylerjereddy** [comment]
Note that I can reproduce the issue on Python 2.7 if I install with --local packages before creating an env there as well. Anyway, I'll carry on that discussion with the conda team I guess.

**Comment by tylerjereddy** [comment]
Ok, I've closed the conda issue -- note that the usage of `export PYTHONNOUSERSITE=True` solves our problem and the default behavior of Python is to include the site packages in the manner we have both observed (see discussion in the conda issue & links therein).

There are probably other ways to exclude the site packages as well so that conda behaves more like virtualenv, which is probably what we want in asv. Without such a setting / suppression of local site packages, there is some risk that running the conda components of the asv testsuite locally will produce different results than in containerized CI tests where there are no local site packages that may pollute the path.

Is this sufficiently worrying to open another issue for asv so that we have robust site package suppression with conda (i.e., matching isolation of the envs for conda and virtualenv)?

**Comment by SylvainCorlay** [comment]
It does not seem that other conda channels than `default`  are taken into account with the version at the tip of master at the moment.

**Comment by pv** [comment]
The tests show otherwise.

**Comment by pv** [comment]
That is, it may be then a difference between the environment at hand,
and what the tests use.

**Comment by SylvainCorlay** [comment]
The tests appear to be setting the `conda_channels` from the Python API (and not getting it from the `conf.json` file).

**Comment by SylvainCorlay** [comment]
Opened #657 to track this specific issue.

**Comment by pv** [comment]
That should not matter. Indeed, it works correctly also in practice for
me when e.g. setting "conda_channels": ["conda-forge"] in the config
file.

Perhaps you have old environments left over --- it doesn't change the
conda_channels automatically?
